File: LeetCode/Study Plan/Practice Problems/Easy/Unique Email Addresses/Cpp/unique_email_addresses.cc
#include <iostream>
#include <vector>
#include <cstring>
#include <unordered_map>

typedef std::vector<std::string> vs;
typedef std::unordered_map<std::string, bool> usb;

int unique_email_addresses(vs email) {
    usb visited;
    for (int i=0; i<email.size(); i++) {
        std::string curr = email[i];
        std::string temp = "";
        bool plus = false;
        bool rate = false;
        int j = 0;
        while(j<curr.length()) {
            if (!rate) {
                if (curr[j] != '@') {
                    if (plus) {
                        j++;
                        continue;
                    }else {
                        if (curr[j] == '+') {
                            j++;
                            plus = true;
                            continue;
                        }else {
                            if (curr[j] != '.') temp+=curr[j];
                        }
                    }
                }else {
                    temp += curr[j];
                    rate = true;
                }
            }else {
                temp += curr[j];
            }
            j++;
        }
        visited[temp] = true;
    }
    return visited.size();
}

int main() {
    std::cout<<unique_email_addresses(vs{"test.email+alex@leetcode.com","test.e.mail+bob.cathy@leetcode.com","testemail+david@lee.tcode.com"})<<"\n";
}

File: LeetCode/Study Plan/Practice Problems/Easy/Unique Email Addresses/Go/unique_email_addresses.go
package main

import (
	"fmt"
	"strings"
)

func unique_email_addresses(email []string) int {
	visited := make(map[string]bool)
	for i := 0; i < len(email); i++ {
		s := email[i]
		split := strings.Split(s, "@")
		split_on_plus := strings.Split(split[0], "+")
		split_on_dot := strings.Split(split_on_plus[0], ".")
		s_final := strings.Join(split_on_dot, "")
		s_final += "@" + split[1]
		visited[s_final] = true
	}
	return len(visited)
}

func main() {
	fmt.Println(unique_email_addresses([]string{"test.email+alex@leetcode.com", "test.e.mail+bob.cathy@leetcode.com", "testemail+david@lee.tcode.com"}))
}
